### PR TITLE
[Core] Add JS APIs for Navbar, NavbarDivider, NavbarGroup, NavbarHeading

### PR DESCRIPTION
**Pull request by cmslewis**
// Second attempt since the whole build change after I opened the last one

#### Checklist

- [x] Include tests
- [x] Update documentation

#### Changes proposed in this pull request:

- 🎉 Added `Navbar` components as simple wrapper around the CSS API:
    - `Navbar`
    - `NavbarDivider` + `Navbar.Divider`
    - `NavbarGroup` + `Navbar.Group`
    - `NavbarHeading` + `Navbar.Heading`
- Added a `Navbar` example:

#### Reviewers should focus on:

- [x] Is this the simplest possible surface area around the CSS API? Any superfluous props we don't need?
- [x] I created empty interfaces for each component's prop definitions for the sake of clarity in the docs. Hope that's okay.
- [x] Do we want a `displayName` for each component?

### CODE REVIEW

File: packages/core/src/components/navbar/navbar.md [hunk 1]
```diff
@@ -40,3 +40,35 @@ Wrap your `.pt-navbar-group`s in an element with your desired `width` and `margi
...
+@## JavaScript API
+
+The `Navbar` component is available in the __@blueprintjs/core__ package.
+Make sure to review the [general usage docs for JS components](#blueprint.usage).
+
+These components are simple wrappers around the corresponding CSS APIs. Each of them supports the full range of HTML props.
+
+```tsx
+<Navbar>
+    <div style="margin: 0 auto; width: 480px;"> <!-- ADD ME -->
```
Comment (giladgray) [inline packages/core/src/components/navbar/navbar.md]: wait why is this is necessary? undocumented awkwardness. 
maybe `<Navbar innerWidth={480}>`?

File: packages/core/src/components/navbar/navbar.md [hunk 2]
```diff
@@ -40,3 +40,35 @@ Wrap your `.pt-navbar-group`s in an element with your desired `width` and `margi
...
+        <NavbarGroup>
+            <NavbarHeading text="Blueprint" />
+        </NavbarGroup>
+        <NavbarGroup align="right">
+            <button class="pt-button pt-minimal pt-icon-home">Home</button>
+            <button class="pt-button pt-minimal pt-icon-document">Files</button>
+            <NavbarDivider />
+            <button class="pt-button pt-minimal pt-icon-user"></button>
+            <button class="pt-button pt-minimal pt-icon-notifications"></button>
+            <button class="pt-button pt-minimal pt-icon-cog"></button>
```
Comment (giladgray) [inline packages/core/src/components/navbar/navbar.md]: use `<Button>` here please

File: packages/core/src/components/navbar/navbarHeading.tsx [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import { IProps } from "../../common/props";
+
+// allow the empty interface so we can label it clearly in the docs
+// tslint:disable-next-line:no-empty-interface
+export interface INavbarHeadingProps extends React.HTMLProps<HTMLDivElement>, IProps {
+    // Empty
+}
+
+// this component is simple enough that tests would be purely tautological.
+/* istanbul ignore next */
```
Comment (giladgray) [inline packages/core/src/components/navbar/navbarHeading.tsx]: 🤔 hmm so we ignore from code coverage but still have simple tests?
are those simple tests really not enough to pass the coverage bar?

File: packages/core/src/components/navbar/navbarHeading.tsx [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+@PureRender
+export class NavbarHeading extends React.Component<React.HTMLProps<HTMLDivElement>, {}> {
+    public static displayName = "Blueprint.NavbarHeading";
+
+    public render() {
+        const { children, className, ...htmlProps } = this.props;
+        return (
+            <div className={classNames(Classes.NAVBAR_HEADING, className)} {...htmlProps}>
+                {children}
```
Comment (giladgray) [inline packages/core/src/components/navbar/navbarHeading.tsx]: _technically_ you could omit `children` from lines 26 and 29 and it would still be set correctly from `{...htmlProps}`, but this is more legible.

**Comment by cmslewis** [comment]
...so much for cherry-picking. Looks like this commit didn't do what it claimed with all the file shuffling: https://github.com/palantir/blueprint/pull/1835/commits/bba6cc6bb103f50377a986c49bb22a9993db41f5

Will fix.

File: packages/core/src/components/navbar/navbar.md (hunk 1 above)
Comment (llorca) [reply packages/core/src/components/navbar/navbar.md]: depends whether you want a flexible width navbar or a fixed width one with full bleed background. I'm inclined to leave this off the API for now, and maybe remove that from the current example

Comment (cmslewis) [reply packages/core/src/components/navbar/navbar.md]: Removed

File: packages/core/src/components/navbar/navbar.md (hunk 2 above)
Comment (cmslewis) [reply packages/core/src/components/navbar/navbar.md]: Had this before. Something went wrong with the cherry-pick.

File: packages/core/src/components/navbar/navbarHeading.tsx (hunk 3 above)
Comment (cmslewis) [reply packages/core/src/components/navbar/navbarHeading.tsx]: No tests should exist. The tests in this PR as of now are also artifacts from the cherry-pick error.

File: packages/core/src/components/navbar/navbarHeading.tsx (hunk 4 above)
Comment (cmslewis) [reply packages/core/src/components/navbar/navbarHeading.tsx]: Ack.

File: packages/core/src/components/navbar/navbar.md [hunk 5]
```diff
@@ -40,3 +40,33 @@ Wrap your `.pt-navbar-group`s in an element with your desired `width` and `margi
 horizontally center it.
 
 @css pt-navbar.pt-container
+
+@## JavaScript API
+
+The `Navbar` component is available in the __@blueprintjs/core__ package.
+Make sure to review the [general usage docs for JS components](#blueprint.usage).
+
+These components are simple wrappers around the corresponding CSS APIs. Each of them supports the full range of HTML props.
```
Comment (giladgray) [inline packages/core/src/components/navbar/navbar.md]: worth mentioning that they can also be used as `Navbar.Group`, etc.

File: packages/core/src/components/navbar/navbar.md [hunk 6]
```diff
@@ -40,3 +40,33 @@ Wrap your `.pt-navbar-group`s in an element with your desired `width` and `margi
...
+    <NavbarGroup>
+        <NavbarHeading text="Blueprint" />
+    </NavbarGroup>
+    <NavbarGroup align="right">
+        <Button class="pt-minimal" iconName="home">Home</Button>
+        <Button class="pt-minimal" iconName="document">Files</Button>
+        <NavbarDivider />
+        <Button class="pt-minimal" iconName="user"></Button>
+        <Button class="pt-minimal" iconName="notifications"></Button>
+        <Button class="pt-minimal" iconName="cog"></Button>
```
Comment (giladgray) [inline packages/core/src/components/navbar/navbar.md]: `className`, throughout this code block

File: packages/docs-app/src/examples/core-examples/navbarExample.tsx [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    public state: INavbarExampleState = {
+        alignRight: false,
+    };
+
+    private handleAlignRightChange = handleBooleanChange(alignRight => this.setState({ alignRight }));
+
+    protected renderExample() {
+        return (
+            <Navbar>
+                <div style={{ margin: "0 auto", width: 480 }}>
```
Comment (giladgray) [inline packages/docs-app/src/examples/core-examples/navbarExample.tsx]: wait so why is this required? it does not appear in the docs code sample.

I think it should be a feature of `Navbar` itself, like `<Navbar innerWidth={480}>`

Comment (cmslewis) [reply packages/docs-app/src/examples/core-examples/navbarExample.tsx]: Fixed via a simple `width: 100%` rule on the wrapping element.

File: packages/core/src/components/navbar/navbarDivider.tsx [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+import * as classNames from "classnames";
+import * as PureRender from "pure-render-decorator";
+import * as React from "react";
+import * as Classes from "../../common/classes";
+import { IProps } from "../../common/props";
+
+// allow the empty interface so we can label it clearly in the docs
+// tslint:disable-next-line:no-empty-interface
+export interface INavbarDividerProps extends IProps {
```
Comment (adidahiya) [inline packages/core/src/components/navbar/navbarDivider.tsx]: you are spreading `...htmlProps` below but not extending from `React.HTMLProps<HTMLDivElement>`. pick one

Comment (cmslewis) [reply packages/core/src/components/navbar/navbarDivider.tsx]: Will fix, but I'll note that this is something a unit test would have caught.

Comment (adidahiya) [reply packages/core/src/components/navbar/navbarDivider.tsx]: sure, but I caught it in CR :)

File: packages/core/src/components/navbar/navbar.md [hunk 9]
```diff
@@ -40,3 +40,37 @@ Wrap your `.pt-navbar-group`s in an element with your desired `width` and `margi
...
+as well: `Navbar.Group`, `Navbar.Heading`, and `Navbar.Divider`, respectively.
+Make sure to review the [general usage docs for JS components](#blueprint.usage).
+
+These components are simple wrappers around the corresponding CSS APIs. Each of
+them supports the full range of HTML props.
+
+```tsx
+<Navbar>
+    <NavbarGroup>
+        <NavbarHeading text="Blueprint" />
```
Comment (giladgray) [inline packages/core/src/components/navbar/navbar.md]: this `text` prop is not defined on the interface. did you mean to use `children`?
